rfd/0174-kube-rbac-bootstrap.md

```diff
@@ -0,0 +1,346 @@
+service KubeProvisionService {
+  // CreateKubeProvision creates a new KubeProvision.
+  rpc CreateKubeProvision(CreateKubeProvisionRequest) returns (KubeProvision);
+  
+  // GetKubeProvision gets a KubeProvision by name.
+  rpc GetKubeProvision(GetKubeProvisionRequest) returns (KubeProvision);
+  
+  // ListKubeProvisions returns a list of KubeProvisions. It supports pagination.
+  rpc ListKubeProvisions(ListKubeProvisionsRequest) returns (ListKubeProvisionsResponse);
+  
+  // UpdateKubeProvision updates an existing KubeProvision.
+  rpc UpdateKubeProvision(UpdateKubeProvisionRequest) returns (KubeProvision);
+  
+  // UpsertKubeProvision upserts a KubeProvision.
+  rpc UpsertKubeProvision(UpsertKubeProvisionRequest) returns (KubeProvision);
+  
+  // DeleteKubeProvision deletes a KubeProvision.
+  rpc DeleteKubeProvision(DeleteKubeProvisionRequest) returns (google.protobuf.Empty);
+}
+```
+
+Every five minutes we will run a reconciliation loop that will compare the current state on Kubernetes clusters under
+Teleport management with the desired state and update the cluster's RBAC if there's a difference. KubeProvision resources
+will be added to the KubeService cache, so we won't need to request potentially large but rarely changed resources from the 
+Auth server every reconciliation iteration. 
+
+As described in the earlier in the RFD, Teleport roles with `kubernetes_permissions` field set will also trigger the reconciliation.
+Relevant Teleport roles will be treated as a "ephemeral KubeProvision" - meaning that we will not actually create KubeProvision for them
+on the backend, but they will reuse same code paths that serves provisioning - Teleport roles will be just another source for the 
+reconciliation process.
+
+Teleport will mark RBAC resources under its control with the "app.kubernetes.io/managed-by: Teleport" label. That way we will
+separate resources managed by Teleport and those managed by the user manually.
+
+Resource labels will be taken into account when doing the reconciliation - users will be able to match different
+Kubernetes clusters for different KubeProvision resources. If a KubeProvision resource doesn't have any labels defined it
+will not match any clusters, effectively being disabled. If a Kubernetes cluster has label `teleport.dev/kube-provision-disabled`, 
+we will exclude this cluster from the provisioning resources, even if some resources will match it.
+
+Existing clusters will not be affected by this feature by default, since there will be not KubeProvision resources to process. User 
+will need to make decision and create KubeProvisions resources to actively start using this feature.
+
+In order to be able to reconcile the state, Teleport will need to perform CRUD operations on Kubernetes RBAC resources.
+When performing CRUD operations, Teleport will impersonate the "system:masters" group for most cases, but on GKE `cluster-admin` 
```

Comment:
This is not always feasible, as customers might block impersonation of `system:masters`. In my opinion, we should ask users to explicitly grant permissions to edit or update roles and cluster roles, rather than randomly impersonating `system:masters`. It's clear and users know what to expect from teleport